File: Data Structures & Algorithms/search-for-word/submission-0.java
class Solution {
    Set<Pair<Integer, Integer>> path = new HashSet<>();
    public boolean exist(char[][] board, String word) {
        for (int r = 0; r < board.length; r++) {
            for (int c = 0; c < board[0].length; c++) {
                if (backtrack(board, word, r, c, 0)) {
                    return true;
                }
            }
        }

        return false;
    }

    public boolean backtrack(char[][] board, String word, int r, int c, int i) {
        if (i == word.length()) {
            return true;
        }

        if (r < 0 || c < 0 || r >= board.length || 
        c >= board[0].length || board[r][c] != word.charAt(i) || 
        path.contains(new Pair<>(r, c))) {
            return false;
        }

        path.add(new Pair<>(r, c));
        boolean res = backtrack(board, word, r + 1, c, i + 1) || 
        backtrack(board, word, r - 1, c, i + 1) || 
        backtrack(board, word, r, c + 1, i + 1) ||
        backtrack(board, word, r, c - 1, i + 1);
        path.remove(new Pair<>(r, c));
        return res;
    }
}

File: Data Structures & Algorithms/palindromic-substrings/submission-0.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        dp = [[False] * len(s) for _ in range(len(s))]
        res = 0

        for length in range(1, len(s) + 1):
            for i in range(len(s) - length + 1):
                j = i + length - 1

                if length == 1:
                    dp[i][j] = True
                elif length == 2:
                    dp[i][j] = s[i] == s[j]
                else:
                    dp[i][j] = s[i] == s[j] and dp[i + 1][j - 1]

                if dp[i][j]:
                    res += 1

        return res